[SimplifyCFG] Treat umul + extract pattern as cheap single instruction.

spaits opened the pull request:
Fix issue #115683 .

Overflow arithmetic instruction plus extract value are usually generated when a division is being replaced, but the zero check may still be there. In that case hoist these two instructions out of this basic block, and let later optimizations take care of the unnecessary zero checks.

An example:
```llvm
define i1 @src(i8 %x, i8 %y) zeroext {
entry:
  %cmp.not = icmp eq i8 %y, 0
  br i1 %cmp.not, label %land.end, label %land.rhs

land.rhs:
  %mul = umul_overflow i8 %y, %x
  %mul.ov = extractvalue {i8, i1} %mul, 1
  br label %land.end

land.end:
  %#0 = phi i1 [ 0, %entry ], [ %mul.ov, %land.rhs ]
  ret i1 %#0
}
=>
define i1 @tgt(i8 %x, i8 %y) zeroext {
entry:
  %fx = freeze i8 %x
  %mul = umul_overflow i8 %y, %fx
  %mul.ov = extractvalue {i8, i1} %mul, 1
  ret i1 %mul.ov
}
```

spaits commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -3290,9 +3290,11 @@ bool SimplifyCFGOpt::speculativelyExecuteBB(BranchInst *BI,
   bool HoistLoadsStores = HoistLoadsStoresWithCondFaulting &&
                           Options.HoistLoadsStoresWithCondFaulting;
   SmallVector<Instruction *, 2> SpeculatedConditionalLoadsStores;
+  InstructionCost BlockCostSoFar = 0;
```
I have introduced this block cost counter for later cases, when we are able to spec. execute more than one instructions. (Does anyone know, why we are limited to one instruction here? If I have some time I may check the git logs more and will try to found out.)

spaits commented:
Also related to:
https://reviews.llvm.org/D65143
And fixes the problem pointed out in:
https://github.com/llvm/llvm-project/commit/6b248fca333831baa0bc3a0de169befd2a832628

goldsteinn commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -3329,20 +3331,29 @@ bool SimplifyCFGOpt::speculativelyExecuteBB(BranchInst *BI,
+    if (match(&I,
+              m_ExtractValue<1>(m_OneUse(
+                  m_Intrinsic<Intrinsic::umul_with_overflow>(m_Value())))) &&
+        ThenBB->size() <= 3)
```
The `ThenBB->size()` includes debug info which we shouldn't be counting.

goldsteinn commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -3329,20 +3331,29 @@ bool SimplifyCFGOpt::speculativelyExecuteBB(BranchInst *BI,
+    // If we don't find any pattern, that must be cheap, then only speculatively
+    // execute a single instruction (not counting the terminator).
+    if (!PatternFound && SpeculatedInstructions > 1)
+      return false;
```
I think it may be easier to just add a `MaxSpeculatedInstructionsToHoist` variable and increment it if you find an overflow intrinsic whose only user is an extract.

goldsteinn replied:
Err, I was thinking something a bit different, particularly I don't really see the reason for accumulating `BlockCostSoFar` for the `extract` and `mul.overflow` if the expressed intent is to treat it as a single instruction.

What about something like:
``` 
    bool PartialInst = false;
    ...
    // Not count load/store into cost if target supports conditional faulting
    // b/c it's cheap to speculate it.
    if (IsSafeCheapLoadStore)
      SpeculatedConditionalLoadsStores.push_back(&I);
    else if (match(&I, m_ExtractValue<1>(
                           m_OneUse(m_WithOverflowInst(m_Value())))) &&
             I.getOperand(0)->getParent() == ThenBB)
      // We are treating (extract (mul/sub/add.overflow), 1) as a single
      // instruction
      PartialInst = true;
    else
      ++SpeculatedInstructions;
     ...
    if (!IsSafeCheapLoadStore && !SpeculatedStoreValue && !PartialInst &&
        computeSpeculationCost(&I, TTI) >
            PHINodeFoldingThreshold * TargetTransformInfo::TCC_Basic)
      return false;
```

Although I'm not really sure if the `PartialInst` part is necessary at all in fact.

spaits replied:
I have removed the accumulating part of the PR. It was not needed. I have just added it because it looked more generic to me(https://github.com/llvm/llvm-project/pull/124933#discussion_r1934099948)(If later we would want to hoist more then one instruction but we would still care about the cost then we would need it, but not now). I know it is a generally bad practice to include code in a commit, that is not related to the main goal of the commit. Sorry for that.

If we don't have the`PartialInst` variable then the hoisting would not be done, because of the high cost.
On rv64, the value of ` PHINodeFoldingThreshold * TargetTransformInfo::TCC_Basic` is 2 and the cost of the pattern is 5. So one of the instructions cost must be 3 and that would be over the threshold.

I was also thinking of extending the pattern. Maybe we should check, if the condition is truly a zero checking for one of the operand of the overflow operator. This hoist is probably beneficial in that case only, because later on the redundant check can be dismissed.
What do you think?

spaits replied:
Also `MaxSpeculatedInstructionsToHoist` is still needed, because the `match` only finds the last instruction of the pattern, the others are still checked and therefore they increase the instruction count.

spaits commented:
@goldsteinn Thank you very much for your review.
I have added the `MaxSpeculatedInstructionsToHoist` variable as you have suggested.
I kept the `PatternFound`. It is used to decide whether we should take in account the calculated cost of the basic block. I could use the `MaxSpeculatedInstructionsToHoist` variable to do this now: if the value of this is one, then there is no cheap patter, else there is. But, what if we would want to hoist more instructions than one, even if there is no patter? For that reason I keep an own variable for patterns.

dtcxzyw reviewed (approved):
LGTM. Thank you! Can you adjust the PR description to make it ready for merge?
Please wait for additional approval from other reviewers :)

nikic commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -3324,26 +3338,36 @@ bool SimplifyCFGOpt::speculativelyExecuteBB(BranchInst *BI,
+    else if (match(&I,
+              m_ExtractValue<1>(m_OneUse(
+                  m_WithOverflowInst(OI))))) {
+      MaxSpeculatedInstructionsToHoist = 3;
```
Why is this 3 and not 2? It's it just with.overflow + extractvalue?

nikic replied:
Also, why don't we increment SpeculatedInstructions in this branch? I think in the end you'll allow 2 more instructions than you actually want. Can you please add a test that shows that you don't actually allow more than just the pattern?

spaits replied:
It was 3 instead of 2 because I have counted in the terminator instruction, but looking again at the loop, it is being dropped, so you are right there is no need for that.

spaits replied:
And when we match the pattern, we are consuming one instruction from the pattern, so you are also right here, we shall increment counter.
Sorry for not paying attention to this earlier.

spaits commented:
@dtcxzyw @nikic Thank you very much for your review.
I have made some changes on the PR description. Is it proper for a commit message?
I have also fixes the pattern length and the instruction counting issue.
Also added a test with some more `add` instructions in the basic block, to see that we are not hoisting, if there are more instructions.
The tests still need a deeper look. I will do more work on them soon. When is use `%add2` in the `%99 = phi i32 ...` the hoisting will happen. This is interesting, because when I do debugging I see the `return false`, when the instruction count hits 3. Maybe these instructions are being hoisted by another part of the SimplifyCFG pass?

nikic commented on llvm/test/Transforms/SimplifyCFG/umul-extract-pattern.ll:
```diff
@@ -0,0 +1,73 @@
+  ret i16 %conv
+}
+
+define dso_local signext range(i32 0, 1024) i32 @noHoist(i64 noundef %x, i64 noundef %y) local_unnamed_addr #0 {
```
Can you please adjust the test so it is the same as the previous one, apart from adding a single extra instruction? Otherwise we aren't testing the exact cut-off point, and other changes between the functions might impact behavior.

spaits replied:
I have added the extra code in `entry` and in `land.end` so the `EphemeralValueTracker` won't track my instruction and we don't have to use the value produced by that extra instruction in the pattern.
Now I use the value calculated by the extra instruction, so the loop won't `continue` because of it.

nikic commented on llvm/test/Transforms/SimplifyCFG/umul-extract-pattern.ll:
```diff
@@ -0,0 +1,73 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt -passes=simplifycfg -simplifycfg-require-and-preserve-domtree=1 -S < %s | FileCheck %s
+
+define dso_local signext range(i16 0, 2) i16 @func2(i64 noundef %x, i64 noundef %y) local_unnamed_addr #0 {
```
You can drop the dso_local, range, etc here.

spaits commented on llvm/test/Transforms/SimplifyCFG/umul-extract-pattern.ll:
```diff
@@ -0,0 +1,54 @@
+  br i1 %cmp.not, label %land.end, label %land.rhs
+
+land.rhs:                                   ; preds = %entry
+  %add = add nsw i64 %y, %x
```
@nikic Even if the extra instruction is here, we still hoist.
What is interesting, if I put the following debug prints in the code when instruction count checking:
```cpp
    if (SpeculatedInstructions > MaxSpeculatedInstructionsToHoist) {
      llvm::errs() << "Returning:\n";
      I.dump();
      llvm::errs() << "\n";
      return false;
    }
    llvm::errs() << "Proceed\n";
    I.dump();
    llvm::errs() << "\n";
```
(I know, not the best way to debug :) )
I get this in my terminal:
```
ninja opt && ../llvm/utils/update_test_checks.py --tool-binary=bin/opt ../llvm/test/Transforms/SimplifyCFG/umul-extract-pattern.ll 
ninja: no work to do.
...
Proceed
  %mul.ov = extractvalue { i64, i1 } %mul, 1

Proceed
  %mul = tail call { i64, i1 } @llvm.umul.with.overflow.i64(i64 %add, i64 %x)

Returning:
  %add = add nsw i64 %y, %x
```
So to me, it seems that we run onto the `return false`.
I will investigate this further.

spaits replied:
Probably it has something to do with my previous observation stated in https://github.com/llvm/llvm-project/pull/124933#issuecomment-2629565199 . In the previous test, the difference between hoisting and not hoisting was the register used in the `phi`. If we used the `add` register and ignored `add2` register (so basically add2's value would just be calculated but never used) we would not hoist. When using `add2` we would hoist.
The debug prints looked the same. I saw the return after reaching the instruction limit in both cases.

spaits committed:
[Transform] Treat umul + extract pattern as cheap single instruction.

spaits commented:
@nikic Sorry for the slow response. I finally had some some time.
The problem with my test was that, I forgot to use the `target triple = "riscv64-unknown-unknown-elf"` attribute for the file.
When this is not there (or any other target) then the hoisting happens without the patch.
The function doing that is the following (If I understand everything correctly):
https://github.com/llvm/llvm-project/blob/838e617069e555468753bf902eb2bfdba3fe6309/llvm/lib/Transforms/Utils/SimplifyCFG.cpp#L8448

Maybe I should look more into that function and try to move this optimization there?

spaits commented:
Gentle ping @nikic .
I have checked `foldTwoEntryPHINode` and `dominatesMergePoint`.

It turns out, that the cost of
```llvm
 %mul = tail call { i64, i1 } @llvm.umul.with.overflow.i64(i64 %add, i64 %x)
```
is 1, when there is no target specified. When using the riscv64 target then this value is 5.

`dominatesMergePoint` works recursively so I think it would be more complex, to "change" the cost of the pattern there. I would personally keep things the way they are in the current version of the patch, but if you would want this to move elsewhere the I am going to do that.

antoniofrighetto commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -3286,7 +3286,21 @@ bool SimplifyCFGOpt::speculativelyExecuteBB(BranchInst *BI,
 
   SmallVector<Instruction *, 4> SpeculatedDbgIntrinsics;
 
+  // The number of already examined instructions. Debug instructions don't
+  // count!
```
```suggestion
  // The number of already examined instructions. Debug instructions don't
  // count.
```

antoniofrighetto commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -3286,7 +3286,21 @@ bool SimplifyCFGOpt::speculativelyExecuteBB(BranchInst *BI,
+  // The number of already examined instructions. Debug instructions don't
+  // count!
   unsigned SpeculatedInstructions = 0;
+  // By default the number of instructions that may be speculatevly executed is
```
```suggestion
  // By default the number of instructions that may be speculatively executed is
```

antoniofrighetto commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -3286,7 +3286,21 @@ bool SimplifyCFGOpt::speculativelyExecuteBB(BranchInst *BI,
+  // count!
   unsigned SpeculatedInstructions = 0;
+  // By default the number of instructions that may be speculatevly executed is
+  // one. Whenever a pattern is found in the basic block, that is cheap for sure
+  // we increase this number to the size of the pattern (how many instructions
+  // are there in that pattern).
```
```suggestion
  // one. Whenever `extract oneuse(op.with.overflow),1` pattern is found,
  // we increase this threshold to the size of the pattern (how many instructions
  // are there in that pattern).
```

antoniofrighetto commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -3286,7 +3286,21 @@ bool SimplifyCFGOpt::speculativelyExecuteBB(BranchInst *BI,
+  // By default the number of instructions that may be speculatevly executed is
+  // one. Whenever a pattern is found in the basic block, that is cheap for sure
+  // we increase this number to the size of the pattern (how many instructions
+  // are there in that pattern).
+  unsigned MaxSpeculatedInstructionsToHoist = 1;
```
Newline before field.

antoniofrighetto commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -3316,34 +3330,45 @@ bool SimplifyCFGOpt::speculativelyExecuteBB(BranchInst *BI,
             PHINodeFoldingThreshold * TargetTransformInfo::TCC_Basic)
       return false;
 
+    // The number of instrcutions to be speculatively executed is limited.
```
```suggestion
    // The number of instructions to be speculatively executed is limited.
```

antoniofrighetto commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -3286,7 +3286,21 @@ bool SimplifyCFGOpt::speculativelyExecuteBB(BranchInst *BI,
+  // only hoisting the cheap pattern only and not other expensive instructions
+  // too, we have the `MaxSpeculatedInstructionsToHoist` variable to track that
+  // the basic block truly only contains that pattern.
+  bool PartialInst = false;
```
Can't reuse MaxSpeculatedInstructionsToHoist instead of having an additional boolean?

spaits replied:
Currently, we could. But if we were to change the number of instructions to speculate from the current value, which is 1 to another value, for example 2, then we would have a problem.
In that case the length of  `MaxSpeculatedInstructionsToHoist` wouldn't necessarily mean, that we have a pattern.

I don't know if there is any plan to change the default value from that, but I was thinking that it is better not to depend on the number of instruction to be hoistable by default.

spaits replied:
We got rid of ` MaxSpeculatedInstructionsToHoist` and now we are only using ` MaxSpeculatedInstructionsToHoist`. I think now this can also be resolved.

spaits committed:
Fix typo

Co-authored-by: Antonio Frighetto <me@antoniofrighetto.com>

spaits committed:
Fix typo

Co-authored-by: Antonio Frighetto <me@antoniofrighetto.com>

spaits committed:
Update llvm/lib/Transforms/Utils/SimplifyCFG.cpp

Co-authored-by: Antonio Frighetto <me@antoniofrighetto.com>

spaits committed:
Update llvm/lib/Transforms/Utils/SimplifyCFG.cpp

Co-authored-by: Antonio Frighetto <me@antoniofrighetto.com>

spaits committed:
Proper spacing for fields

spaits commented:
@antoniofrighetto Thank you very much for your review.
I think I have addressed all the reviews. Sorry for the amount of typos.
In the case of your last comment about ` MaxSpeculatedInstructionsToHoist` I think we should wait for other reviewers opinions too.

antoniofrighetto commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -3316,34 +3333,45 @@ bool SimplifyCFGOpt::speculativelyExecuteBB(BranchInst *BI,
 
+    // The number of instructions to be speculatively executed is limited.
+    // This limit is dependent on the found patterns.
+    if (SpeculatedInstructions > MaxSpeculatedInstructionsToHoist)
+      return false;
```
It may feel like we'd need more of a cl::opt here, if we were to maintain a threshold of this kind. I wonder if dropping MaxSpeculatedInstructionsToHoist and have this doesn't look too contrived for now:
```cpp
if (SpeculatedInstructions > (PartialInst ? 2 : 1))
  return false;
```

spaits replied:
I have dropped `MaxSpeculatedInstructionsToHoist`.

spaits committed:
Get rid of MaxSpeculatedInstructionsToHoist

antoniofrighetto commented:
@spaits Looks like the CI is reporting a failure in PhaseOrdering/X86/unsigned-multiply-overflow-check.ll test, which should require an update, could you look at it please?

spaits committed:
Fix failing test case

spaits commented:
@antoniofrighetto I have fixed the failing test case. It seems like we have a regression, compared to some earlier state of the PR, that has been lost to squashing. Now we get back the original state of `llvm/test/Transforms/PhaseOrdering/X86/unsigned-multiply-overflow-check.ll`. I will check why it no longer fixes that X86 test case.

spaits commented:
So, the only reason it has worked before was the omission of the checking how many instructions will be hoisted.

spaits commented:
I think I will try to move this optimization to to `foldTwoEntryPHINode`. That function is calling the `dominatesMergePoint` function. In that I will match to the pattern and set a low cost value for this pattern.

antoniofrighetto commented:
Perhaps we could include `!extract oneuse(op.with.overflow),1` in the pattern, have something like (and adjust the limit):
```cpp
WithOverflowInst *UnusedWO;
if (match(&I, m_CombineOr(m_ExtractValue<1>(m_OneUse(m_WithOverflowInst(UnusedWO))),
                          m_Not(m_ExtractValue<1>(m_OneUse(m_WithOverflowInst(UnusedWO)))))))
  Found = true;
```
But TBH, I believe the current code should be fine (regardless of the test).

spaits commented:
I have created another solution for this problem. It is in #128021. I think that is a better one. Please check that too.
